### ISSUE TITLE
Enable passing result from *_step Model Hook to corresponding *_batch_end Callback

### ISSUE THREAD

**Issue by romesco** [PyTorchLightning/pytorch-lightning#3249]
## 🚀 Feature
Give users the option to pass a result at the end of `*_step` directly to the corresponding `*_batch_end` `Callback`.

Example:
`validation_step` outputs prediction and probabilities. Pass these to `validation_batch_end` user-defined `Callback` for advanced logging.

### Motivation

This will remove the need for some boilerplate in the model code and also remove the need for calling the model's `forward()` on the batch inside the callback just to do logging.

**Comment by romesco** [comment]
I can't do this myself, but please assign to @romesco and @williamFalcon !

**Comment by edenlightning** [comment]
@romesco want to try and submit a PR for this?

**Comment by romesco** [comment]
Yep! I started it I'll put out the draft PR tomorrow or so. Got caught pushing for a deadline so some of my open source stuff got backlogged =].
